src/AIAPI/MainGame/GameExtensions.cs

```diff
@@ -0,0 +1,146 @@
+
+            var agentTable = Manager.Map.Instance.AgentTable;            
+            if (agentTable == null || agentTable.Count <= 0) return null;
+
+            foreach (int key in agentTable.Keys)
+			{
+				if (agentTable.TryGetValue(key, out AgentActor agentActor))
+				{
+                    if (agentActor?.charaID == agentData.param.charaID) return agentActor;
+                }
+            }  
+
+            return null;
+        }
+
+        /// <summary>
+        /// Get ChaFiles that are related to this AgentData (heroine). Warning: It might not return some copies.
+        /// </summary>
+        public static IEnumerable<ChaFileControl> GetRelatedChaFiles(this AgentData agentData)
+        {
+            if (agentData == null) throw new ArgumentNullException(nameof(agentData));
+
+            var results = new HashSet<ChaFileControl>();
+
+            if (agentData.param?.actor?.ChaControl?.chaFile != null)
+                results.Add(agentData.param.actor.ChaControl.chaFile);
+
+            var agentActor = agentData.GetNPC();
+            if (agentActor?.ChaControl?.chaFile != null)
+                results.Add(agentActor.ChaControl.chaFile);
+
+            return results;
+        }
+
+        /// <summary>
+        /// Get the persisting player object that describes this character.
+        /// Returns null if the player could not be found. Works only in the main game.
+        /// </summary>
+        public static PlayerActor GetPlayer(this ChaControl chaControl)
+        {
+            if (chaControl == null) throw new ArgumentNullException(nameof(chaControl));
+
+            if (!Manager.Map.IsInstance() || Manager.Map.Instance.Player == null) return null;
+            return Manager.Map.Instance.Player.ChaControl == chaControl ? Manager.Map.Instance.Player : null;
+        }
+
+        /// <summary>
+        /// Get the persisting player object that describes this character.
+        /// Returns null if the player could not be found. Works only in the main game.
+        /// </summary>
+        public static PlayerActor GetPlayer(this ChaFileControl chaFile)
+        {
+            if (chaFile == null) throw new ArgumentNullException(nameof(chaFile));
+
+            if (!Manager.Map.IsInstance() || Manager.Map.Instance.Player == null) return null;
+            return Manager.Map.Instance.Player.GetRelatedChaFiles().Contains(chaFile)
+                ? Manager.Map.Instance.Player
+                : null;
+        }
+
+        /// <summary>
+        /// Get ChaFiles that are related to this player. Warning: It might not return some copies.
+        /// </summary>
+        public static IEnumerable<ChaFileControl> GetRelatedChaFiles(this PlayerActor player)
+        {
+            if (player == null) throw new ArgumentNullException(nameof(player));
+
+            var results = new HashSet<ChaFileControl>();
+
+            if (player.ChaControl?.chaFile != null)
+                results.Add(player.ChaControl.chaFile);
```

Comment:
In KK this returns multiple chaFiles. But PlayerActor only appears to have one chaFile, so it will only ever return one, unless I missed something.

Reply:
This is a quirk of how KK main game copies characters around, it's a hack to keep all different copies up to date.